File: 14-TypeScript/Grey Multi-Tenant/packages/grey-zig/src/config/options.zig
//! Grey SDK Configuration Options
//!
//! Configuration for the Grey SDK client.

const std = @import("std");

/// Configuration options for the Grey SDK client.
pub const Options = struct {
    /// API host
    host: []const u8 = "localhost",
    /// API port
    port: u16 = 50051,
    /// Whether to use TLS
    use_tls: bool = false,
    /// Request timeout in milliseconds
    timeout_ms: u32 = 30000,
    /// Authentication token
    auth_token: ?[]const u8 = null,
    /// Custom metadata headers
    metadata: ?std.StringHashMap([]const u8) = null,
    /// Allocator for dynamic data
    allocator: std.mem.Allocator,

    const Self = @This();

    /// Create default options for local development
    pub fn local(allocator: std.mem.Allocator) Self {
        return .{ .allocator = allocator };
    }

    /// Create options for local development with custom port
    pub fn localWithPort(allocator: std.mem.Allocator, port: u16) Self {
        return .{
            .allocator = allocator,
            .port = port,
        };
    }

    /// Create options for production
    pub fn production(allocator: std.mem.Allocator, host: []const u8) Self {
        return .{
            .allocator = allocator,
            .host = host,
            .port = 443,
            .use_tls = true,
        };
    }

    /// Create options for production with custom port
    pub fn productionWithPort(allocator: std.mem.Allocator, host: []const u8, port: u16) Self {
        return .{
            .allocator = allocator,
            .host = host,
            .port = port,
            .use_tls = true,
        };
    }

    /// Get the gRPC endpoint string
    pub fn endpoint(self: Self) ![]u8 {
        return std.fmt.allocPrint(self.allocator, "{s}:{d}", .{ self.host, self.port });
    }

    /// Create a new options with auth token set
    pub fn withAuthToken(self: Self, token: []const u8) Self {
        var new = self;
        new.auth_token = token;
        return new;
    }

    /// Create a new options with timeout set
    pub fn withTimeout(self: Self, timeout_ms: u32) Self {
        var new = self;
        new.timeout_ms = timeout_ms;
        return new;
    }

    /// Add a metadata header
    pub fn withMetadata(self: *Self, key: []const u8, value: []const u8) !void {
        if (self.metadata == null) {
            self.metadata = std.StringHashMap([]const u8).init(self.allocator);
        }
        try self.metadata.?.put(key, value);
    }

    /// Clean up resources
    pub fn deinit(self: *Self) void {
        if (self.metadata) |*m| {
            m.deinit();
        }
    }
};

test "Options local" {
    var options = Options.local(std.testing.allocator);
    defer options.deinit();

    try std.testing.expectEqualStrings("localhost", options.host);
    try std.testing.expectEqual(@as(u16, 50051), options.port);
    try std.testing.expectEqual(false, options.use_tls);
}

test "Options production" {
    var options = Options.production(std.testing.allocator, "api.grey.io");
    defer options.deinit();

    try std.testing.expectEqualStrings("api.grey.io", options.host);
    try std.testing.expectEqual(@as(u16, 443), options.port);
    try std.testing.expectEqual(true, options.use_tls);
}

test "Options endpoint" {
    var options = Options.local(std.testing.allocator);
    defer options.deinit();

    const ep = try options.endpoint();
    defer options.allocator.free(ep);

    try std.testing.expectEqualStrings("localhost:50051", ep);
}

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-haskell/src/Grey/Config/Options.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}

-- |
-- Module      : Grey.Config.Options
-- Description : Configuration options for the Grey SDK
-- License     : MIT
--
-- Provides configuration options for connecting to the Grey API.

module Grey.Config.Options
  ( -- * Options Type
    Options(..)
    -- * Constructors
  , defaultOptions
  , localOptions
  , productionOptions
    -- * Modifiers
  , withHost
  , withPort
  , withTls
  , withTimeout
  , withAuthToken
  , withHeader
    -- * Utilities
  , getEndpoint
  ) where

import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Text (Text)
import qualified Data.Text as T
import GHC.Generics (Generic)

-- | Configuration options for the Grey SDK client.
data Options = Options
  { optHost          :: !Text              -- ^ API host
  , optPort          :: !Int               -- ^ API port
  , optUseTls        :: !Bool              -- ^ Whether to use TLS
  , optTimeoutSecs   :: !Int               -- ^ Request timeout in seconds
  , optAuthToken     :: !(Maybe Text)      -- ^ Authentication token
  , optCustomHeaders :: !(Map Text Text)   -- ^ Custom headers
  } deriving (Eq, Show, Generic)

-- | Default options for local development.
defaultOptions :: Options
defaultOptions = Options
  { optHost = "localhost"
  , optPort = 50051
  , optUseTls = False
  , optTimeoutSecs = 30
  , optAuthToken = Nothing
  , optCustomHeaders = Map.empty
  }

-- | Options for local development with optional port.
localOptions :: Maybe Int -> Options
localOptions mPort = defaultOptions
  { optPort = maybe 50051 id mPort
  }

-- | Options for production with host and optional port.
productionOptions :: Text -> Maybe Int -> Options
productionOptions host mPort = Options
  { optHost = host
  , optPort = maybe 443 id mPort
  , optUseTls = True
  , optTimeoutSecs = 30
  , optAuthToken = Nothing
  , optCustomHeaders = Map.empty
  }

-- | Set the host.
withHost :: Text -> Options -> Options
withHost host opts = opts { optHost = host }

-- | Set the port.
withPort :: Int -> Options -> Options
withPort port opts = opts { optPort = port }

-- | Set whether to use TLS.
withTls :: Bool -> Options -> Options
withTls useTls opts = opts { optUseTls = useTls }

-- | Set the timeout in seconds.
withTimeout :: Int -> Options -> Options
withTimeout secs opts = opts { optTimeoutSecs = secs }

-- | Set the authentication token.
withAuthToken :: Text -> Options -> Options
withAuthToken token opts = opts { optAuthToken = Just token }

-- | Add a custom header.
withHeader :: Text -> Text -> Options -> Options
withHeader name value opts = opts
  { optCustomHeaders = Map.insert name value (optCustomHeaders opts)
  }

-- | Get the gRPC endpoint string.
getEndpoint :: Options -> Text
getEndpoint opts =
  optHost opts <> ":" <> T.pack (show (optPort opts))
